torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp

```diff
@@ -1405,6 +1405,34 @@ ProcessGroupNCCL::Options::Options(bool is_high_priority_stream)
     : ProcessGroup::Options(NCCL_BACKEND_NAME),
       is_high_priority_stream(is_high_priority_stream) {}
 
+void ProcessGroupNCCL::startCoalescing() {
+  coalescedDevices_.clear();
+  coalescing_active_ = true;
+  groupStart();
+}
+
+void ProcessGroupNCCL::endCoalescing(
+    std::vector<c10::intrusive_ptr<ProcessGroup::Work>>& reqs) {
+  groupEnd();
+  if (reqs.size() != coalescedDevices_.size()) {
+    TORCH_CHECK(false, "Number of requests do not match number of collectives");
+  }
+
+  int batch_idx = 0;
+  for (const auto& req : reqs) {
+    auto ncclWork = dynamic_cast<ProcessGroupNCCL::WorkNCCL*>(req.get());
```

Comment:
can we static_cast<> here so we have compile-time guarantees and don't blow up at runtime?